### PR TITLE
Optimize `TSDataset.describe` and `TSDataset.info` by vectorization

### LINKED ISSUE #1341: Optimize `TSDataset.describe`
### 🚀 Feature Request

We should optimize method `TSDataset.describe`, because it can consume up to 30% of all computation time during backtest on `NaiveModel` with 10k segments.

### Proposal

In current implementation the bottleneck is `TSDataset._gather_segments_data` and it should be optimized. The problem lies in per-segment iteration.

Possible solution:
- Vectorization
- Optimization of one iteration
- Rewriting cycle using `numba`

As an alternative we could optimize the places where `TSDataset.describe` is used:
- [`BasePipeline._make_predict_timestamps`](https://github.com/tinkoff-ai/etna/blob/cceb50021b7a25bc58efe3fc7c4a8d646c05b309/etna/pipeline/base.py#L457)
- [`FoldMask.validate_on_dataset`](https://github.com/tinkoff-ai/etna/blob/cceb50021b7a25bc58efe3fc7c4a8d646c05b309/etna/pipeline/base.py#L99)
- Check other places

### Test cases

Make sure current tests pass.

### Additional context

Connected issues: #1336.

### PR DESCRIPTION
**Pull request by Mr-Geekman**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Proposed Changes
Look at #1341.

## Closing issues
Closes #1341.

### CODE REVIEW

**Comment by Mr-Geekman** [comment]
Script for testing `describe`:
```python
import time
import json

import numpy as np
import pandas as pd
from loguru import logger

from etna.models import NaiveModel
from etna.datasets import TSDataset, generate_ar_df
from etna.metrics import MAE
from etna.pipeline import Pipeline

HORIZON = 14


def make_df(num_segments: int, num_features: int, num_periods: int, random_state: int = 0) -> pd.DataFrame:
    rng = np.random.default_rng(random_state)
    df = generate_ar_df(
        periods=num_periods, start_time="2020-01-01", n_segments=num_segments
    )

    for i in range(num_features):
        # add int column
        df[f"new_int_{i}"] = rng.integers(low=-100, high=100, size=df.shape[0])

    return df


def check_time(num_segments: int, num_features: int, num_periods: int = 365):
    df = make_df(num_segments=num_segments, num_features=num_features, num_periods=num_periods)
    df_wide = TSDataset.to_dataset(df)
    ts = TSDataset(df=df_wide, freq="D")

    start_time = time.perf_counter()
    _ = ts.describe()
    elapsed_time = time.perf_counter() - start_time

    return elapsed_time


def main():
    num_segments = [10, 100, 1000, 10_000, 100_000]
    num_features = [0, 3, 10]

    results = []
    for cur_num_segments in num_segments:
        for cur_num_features in num_features:
            time_result = check_time(num_segments=cur_num_segments, num_features=cur_num_features)
            record = {"num_segments": cur_num_segments, "num_features": cur_num_features, "time": time_result}
            results.append(record)
            logger.info(json.dumps(record))

    json.dump(results, open("records.json", "w"), indent=2)

if __name__ == "__main__":
    main()
```

Results before optimization:
```json
[
  {
    "num_segments": 10,
    "num_features": 0,
    "time": 0.007862442000000414
  },
  {
    "num_segments": 10,
    "num_features": 3,
    "time": 0.00775050900000096
  },
  {
    "num_segments": 10,
    "num_features": 10,
    "time": 0.00862645999999856
  },
  {
    "num_segments": 100,
    "num_features": 0,
    "time": 0.06804819900000147
  },
  {
    "num_segments": 100,
    "num_features": 3,
    "time": 0.05528060099999976
  },
  {
    "num_segments": 100,
    "num_features": 10,
    "time": 0.05490351599999954
  },
  {
    "num_segments": 1000,
    "num_features": 0,
    "time": 0.511956906
  },
  {
    "num_segments": 1000,
    "num_features": 3,
    "time": 0.5077033259999997
  },
  {
    "num_segments": 1000,
    "num_features": 10,
    "time": 0.49680727800000035
  },
  {
    "num_segments": 10000,
    "num_features": 0,
    "time": 5.198245515000002
  },
  {
    "num_segments": 10000,
    "num_features": 3,
    "time": 5.023976880999999
  },
  {
    "num_segments": 10000,
    "num_features": 10,
    "time": 5.116792693999997
  },
  {
    "num_segments": 100000,
    "num_features": 0,
    "time": 50.777624478999996
  },
  {
    "num_segments": 100000,
    "num_features": 3,
    "time": 51.87359783100001
  },
  {
    "num_segments": 100000,
    "num_features": 10,
    "time": 62.32446584499996
  }
]
```

Results after optimization:
```json
[
  {
    "num_segments": 10,
    "num_features": 0,
    "time": 0.006445242999999934
  },
  {
    "num_segments": 10,
    "num_features": 3,
    "time": 0.005044411000000082
  },
  {
    "num_segments": 10,
    "num_features": 10,
    "time": 0.00412322800000009
  },
  {
    "num_segments": 100,
    "num_features": 0,
    "time": 0.007066482000000818
  },
  {
    "num_segments": 100,
    "num_features": 3,
    "time": 0.006966671999999008
  },
  {
    "num_segments": 100,
    "num_features": 10,
    "time": 0.006875658000000229
  },
  {
    "num_segments": 1000,
    "num_features": 0,
    "time": 0.015869262000000717
  },
  {
    "num_segments": 1000,
    "num_features": 3,
    "time": 0.018922749999999766
  },
  {
    "num_segments": 1000,
    "num_features": 10,
    "time": 0.019158535000000754
  },
  {
    "num_segments": 10000,
    "num_features": 0,
    "time": 0.05820048300000025
  },
  {
    "num_segments": 10000,
    "num_features": 3,
    "time": 0.07253477299999922
  },
  {
    "num_segments": 10000,
    "num_features": 10,
    "time": 0.0792398090000006
  },
  {
    "num_segments": 100000,
    "num_features": 0,
    "time": 0.47934153599999973
  },
  {
    "num_segments": 100000,
    "num_features": 3,
    "time": 0.6095439070000026
  },
  {
    "num_segments": 100000,
    "num_features": 10,
    "time": 0.9615795119999859
  }
]
```

**Comment by Mr-Geekman** [comment]
Results for script from #1338:
```json
[
  {
    "num_segments": 10,
    "num_features": 0,
    "time": 0.18644669199999964
  },
  {
    "num_segments": 10,
    "num_features": 3,
    "time": 0.44299768199999967
  },
  {
    "num_segments": 10,
    "num_features": 10,
    "time": 0.3184416309999998
  },
  {
    "num_segments": 100,
    "num_features": 0,
    "time": 0.40837533100000023
  },
  {
    "num_segments": 100,
    "num_features": 3,
    "time": 0.49695419899999926
  },
  {
    "num_segments": 100,
    "num_features": 10,
    "time": 0.6302544880000003
  },
  {
    "num_segments": 1000,
    "num_features": 0,
    "time": 2.3735116309999995
  },
  {
    "num_segments": 1000,
    "num_features": 3,
    "time": 2.3557946890000014
  },
  {
    "num_segments": 1000,
    "num_features": 10,
    "time": 3.484642255999999
  },
  {
    "num_segments": 10000,
    "num_features": 0,
    "time": 18.414344812
  },
  {
    "num_segments": 10000,
    "num_features": 3,
    "time": 23.947836302000006
  },
  {
    "num_segments": 10000,
    "num_features": 10,
    "time": 37.481222474999996
  }
]
```

File: etna/datasets/tsdataset.py [hunk 1]
```diff
@@ -1320,25 +1320,35 @@ def _gather_common_data(self) -> Dict[str, Any]:
...
+        size = df.shape[0]
+        not_na = ~np.isnan(df.values)
+        min_idx = np.argmax(not_na, axis=0)
+        max_idx = size - np.argmax(not_na[::-1, :], axis=0) - 1
+
+        segments_dict = {}
+        segments_dict["start_timestamp"] = df.index[min_idx].to_series(index=segments)
+        segments_dict["end_timestamp"] = df.index[max_idx].to_series(index=segments)
+        size_borders = min_idx + (size - max_idx - 1)
+        segments_dict["length"] = pd.Series(size - size_borders, dtype="Int64", index=segments)
```
Comment (alex-hse-repository) [inline etna/datasets/tsdataset.py:1343]: May be better:
```
series_len = max_idx - min_idx + 1
num_missing = series_len - np.sum(not_na, axis=0) # Max num not nans - Num not nans
```

File: etna/datasets/tsdataset.py [hunk 2]
```diff
@@ -1320,25 +1320,35 @@ def _gather_common_data(self) -> Dict[str, Any]:
...
+        if segments is None:
+            segments_slice = slice(None)
+            segments = self.segments
+        else:
+            segments_slice = segments
+            segments = segments
+
+        df = self.df.loc[:, (segments_slice, "target")]
+
+        size = df.shape[0]
```
Comment (alex-hse-repository) [inline etna/datasets/tsdataset.py:1334]: `size` is not the best naming, may be num_timestamps or sth like this

File: tests/test_datasets/test_dataset.py [hunk 3]
```diff
@@ -848,61 +873,56 @@ def test_fit_transform_raise_warning_on_diff_endings(ts_diff_endings):
...
-    df.iloc[-3:, 1] = np.NaN
-    ts = TSDataset(df=df, df_exog=df_exog, freq="D", known_future=known_future)
-    segments = ts.segments
-    segments_dict = ts._gather_segments_data(segments)
-    segment_df = pd.DataFrame(segments_dict, index=segments)
+    segments_dict = ts_info._gather_segments_data(ts_info.segments)
+    segment_df = pd.DataFrame(segments_dict, index=ts_info.segments)
 
-    assert np.all(segment_df.index == ts.segments)
+    assert np.all(segment_df.index == ts_info.segments)
```
Comment (alex-hse-repository) [inline tests/test_datasets/test_dataset.py:893]: Looks strange as you created dataframe with this index in the previous line

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]
Nice speed up))
What is the bottleneck in the backtest metrics computation now?

**Review by alex-hse-repository — APPROVED** [review]

**Pull request merged**